### DISCUSSION TITLE
ERC20 OutToken Make Deploy Error

### DISCUSSION THREAD

**Discussion by ArlisVilipold** [Cyfrin/foundry-full-course-cu#2150] [Q&A]
When i try to Deploy OurToken.sol it gives an error. 

error: unexpected argument '--broadcast' found

  tip: to pass '--broadcast' as a value, use '-- --broadcast'

Usage: forge create <--rpc-url <URL>|--flashbots|--jwt-secret <JWT_SECRET>> <--interactive|--private-key <RAW_PRIVATE_KEY>|--mnemonic <MNEMONIC>|--mnemonic-passphrase <PASSPHRASE>|--mnemonic-derivation-path <PATH>|--mnemonic-index <INDEX>> <CONTRACT>

**Comment by EngrPips** [comment]
Hello @ArlisVilipold, what command did you run to deploy the `OurToken` contract?

**Reply by ArlisVilipold** [reply]
make deploy

**Reply by EngrPips** [reply]
What does this command point to in your `Makefile`?

**Reply by ArlisVilipold** [reply]
Makefile is from github repo. 


deploy:
	@forge create src/OurToken.sol:OurToken --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast



-include .env

.PHONY: all test test-zk clean deploy fund help install snapshot format anvil install deploy deploy-zk deploy-zk-sepolia deploy-sepolia verify

DEFAULT_ANVIL_KEY := 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
DEFAULT_ZKSYNC_LOCAL_KEY := 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110

all: clean remove install update build

# Clean the repo
clean  :; forge clean

# Remove modules
remove :; rm -rf .gitmodules && rm -rf .git/modules/* && rm -rf lib && touch .gitmodules && git add . && git commit -m "modules"

install :; forge install cyfrin/foundry-devops@0.2.2 --no-commit && forge install foundry-rs/forge-std@v1.8.2 --no-commit && forge install openzeppelin/openzeppelin-contracts@v5.0.2 --no-commit

# Update Dependencies
update:; forge update

build:; forge build

test :; forge test 

test-zk :; foundryup-zksync && forge test --zksync && foundryup

snapshot :; forge snapshot

format :; forge fmt

anvil :; anvil -m 'test test test test test test test test test test test junk' --steps-tracing --block-time 1

deploy:
	@forge create src/OurToken.sol:OurToken --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast

deploy-sepolia:
	@forge script script/DeployOurToken.s.sol:DeployOurToken --rpc-url $(SEPOLIA_RPC_URL) --account $(ACCOUNT) --sender $(SENDER) --etherscan-api-key $(ETHERSCAN_API_KEY) --broadcast --verify

deploy-zk:
	@forge create src/OurToken.sol:OurToken --rpc-url http://127.0.0.1:8011 --private-key $(DEFAULT_ZKSYNC_LOCAL_KEY) --legacy --zksync

deploy-zk-sepolia:
	@forge create src/OurToken.sol:OurToken --rpc-url $(ZKSYNC_SEPOLIA_RPC_URL) --account $(ACCOUNT) --legacy --zksync

deploy-zk-bad:
	@forge create src/OurToken.sol:OurToken --rpc-url https://sepolia.era.zksync.dev --private-key $(PRIVATE_KEY) --legacy --zksync

verify:
	@forge verify-contract --chain-id 11155111 --num-of-optimizations 200 --watch --constructor-args 0x00000000000000000000000000000000000000000000d3c21bcecceda1000000 --etherscan-api-key $(ETHERSCAN_API_KEY) --compiler-version v0.8.19+commit.7dd6d404 0x089dc24123e0a27d44282a1ccc2fd815989e3300 src/OurToken.sol:OurToken

**Reply by EngrPips** [reply]
Okay
```js
deploy:
https://github.com/forge create src/OurToken.sol:OurToken --rpc-url http://localhost:8545/ --private-key $(DEFAULT_ANVIL_KEY) --broadcast
```
Remove the broadcast flag from this and retry. Let's see how it will play out.

**Reply by philbugcatcher** [reply]
Hello! I ran into the same error. I did as you suggested, and got this error below. Any idea on how I can fix it?

```
[⠊] Compiling...
[⠒] Compiling 6 files with Solc 0.8.26
[⠘] Installing Solc version 0.8.26
[⠘] Successfully installed Solc 0.8.26
[⠃] Solc 0.8.26 finished in 4.68s
Compiler run successful!
Error: 
no arguments provided for contract constructor; consider --constructor-args or --constructor-args-path

Context:
- Error #0: failed to deploy contract
- Error #1: encode length mismatch: expected 1 types, got 0
make: *** [deploy] Error 1
```

**Reply by EngrPips** [reply]
Hello @ArlisVilipold, Does your contract have a constructor that is expecting an argument, and are you following Patrick or you are doing this yourself? 

`no arguments provided for contract constructor; consider --constructor-args or --constructor-args-path`

**Reply by philbugcatcher** [reply]
It says that no arguments were provided for the contract constructor, however I think I did provide the arguments for both constructors used, as below:

## OurToken.sol
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract OurToken is ERC20 {
    constructor(uint256 initialSupply) ERC20("Our Token", "OT") {
        _mint(msg.sender, initialSupply);
    }
}
```

## DeployOurToken.sol
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {OurToken} from "src/OurToken.sol";

contract DeployOurToken is Script {
    uint256 public constant INITIAL_SUPPLY = 1000 ether;

    function run() external {
        vm.startBroadcast();
        new OurToken(INITIAL_SUPPLY);
        vm.stopBroadcast();
    }
}
```

**Reply by philbugcatcher** [reply]
@EngrPips following Patrick

**Reply by EngrPips** [reply]
The below is what my `Makefile` looks like 

```js
-include .env

.PHONY: all test deploy

help: 
	@echo "Useage"
	@echo " make deploy [ARGS=...]"

install:; forge install cyfrin/foundry-devops --no-commit && forge install foundry-rs/forge-std --no-commit && forge install transmissions11/solmate --no-commit && forge install smartcontractkit/chainlink-brownie-contracts@0.6.1 --no-commit

build:; forge build

FUNCTION_TO_TEST ?= test_transferWorksAppropriately

test:; forge test

testUnit:; forge test --mt $(FUNCTION_TO_TEST)




NETWORK_ARGS := --rpc-url http://localhost:8545 --private-key $(PRIVATE_KEY) --broadcast

# if --network sepolia is used then use sepolia details in deployment but if not use anvil details
ifeq ($(findstring --network sepolia,$(ARGS)),--network sepolia)
	NETWORK_ARGS := --rpc-url $(SEPOLIA_RPC_URL) --private-key $(SEPOLIA_PRIVATE_KEY) --broadcast --verify --etherscan-api-key $(ETHERSCAN_API_KEY) -vvvv
endif

anvil: 
	anvil -m 'test test test test test test test test test test test test junk' --steps-tracing --block-time 1

deploy:
	forge script script/DeployMyToken.s.sol:DeployMyToken $(NETWORK_ARGS)

deploy-sepolia:
	forge script script/DeployFundMe.s.sol:DeployFundMe --rpc-url $(SEPOLIA_RPC_URL) --private-key $(SEPOLIA_PRIVATE_KEY) --broadcast --verify --etherscan-api-key $(ETHERSCAN_API_KEY) -vvvv
```

If you want to deploy your contract through a deployment script then the Forge command that you will run should look like the below 

```js
deploy:
	forge script script/DeployMyToken.s.sol:DeployMyToken $(NETWORK_ARGS)
```

**Reply by EngrPips** [reply]
> @EngrPips following Patrick

So, did Patrick run into this issue that you are running into?

**Reply by ArlisVilipold** [reply]
I'm following Patrick, and he did not ran into this error.

**Reply by philbugcatcher** [reply]
@EngrPips 

No, his contract compiled and deployed just fine.

But I noticed that on the Makefile he suggested using ([here](https://github.com/Cyfrin/foundry-erc20-cu/blob/main/Makefile)), the deploy is deploying the contract, and not the script - see below. I think this is why it was not sending the constructor

```
deploy:
	@forge create src/OurToken.sol:OurToken --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast
```

Thank you very much for your help

**Reply by philbugcatcher** [reply]
@ArlisVilipold apologies for hijacking your post. Changing the deploy on the Makefile for this one worked for me:

```
deploy:
	@forge script script/DeployOurToken.s.sol:DeployOurToken --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast
```

**Reply by ArlisVilipold** [reply]
@felliperfm Thank you, it worked for me aswell .

**Reply by EngrPips** [reply]
Awesome, I am glad you figured it out @felliperfm and  @ArlisVilipold. Keep crushing it guys!

**Reply by smilingkylan** [reply]
Thank you for the help @philmnds. That worked for me!

**Reply by EngrPips** [reply]
You guys should keep crushing it.

**Reply by operation-c** [reply]
An additional work around could be the following:
```
forge create src/OurToken.sol:OurToken --rpc-url http://localhost:8545 --account DEFAULT_ANVIL_KEY --constructor-args 1000000
```
assuming you used cast via  `--interactive`

**Reply by EngrPips** [reply]
Interesting

**Reply by suddenly1990** [reply]
> @ArlisVilipold apologies for hijacking your post. Changing the deploy on the Makefile for this one worked for me:为劫持您的帖子道歉。更改 Makefile 上的部署对我有用：
> 
> ```
> deploy:
> 	@forge script script/DeployOurToken.s.sol:DeployOurToken --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast
> ```

 Thank you very much! I am back to watch the video, and after watching it, I noticed that it is different from the GitHub Makefile. The image is a snapshot from the video.
![图像2024-9-2 21 30](https://github.com/user-attachments/assets/dc43a6e4-55c6-47d8-bb1e-bf26ab373b12)

**Reply by EngrPips** [reply]
`$(NETWORK_ARGS)` is equals `--rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast` and the `$(NETWORK_ARGS)` will be changing base on the chain you are running your process on.

**Reply by Totiruzi** [reply]
@EngrPips  @operation-c Please do you know why when I run the command "make deploy"

I get the error
```error
error: unexpected argument '--broadcast' found
```
I copied the Makefile from the project repository for project ourToken.

When I copied the command and ran it without the --broadcast flag I go an error

```Error
no arguments provided for contract constructor; consider --constructor-args or --constructor-args-path
```
I saw above someone suggested passing a value to the --constructor-args. which I did and everything compiled well.

Am I missing something?
Thanks )

**Reply by EngrPips** [reply]
When you pass a value to the constructor args, everything compiles fine, right?

**Reply by EngrPips** [reply]
what does the `make deploy` command point to in your `Makefile`?

**Reply by Totiruzi** [reply]
Yes when I pass a constructor value, everything compiles.

```Makefile

deploy:
	@forge create src/MyToken.sol:MyToken --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast

```

**Reply by EngrPips** [reply]
Remove the `--broadcast` flag and retry let's see how that play out.

**Reply by operation-c** [reply]
@Totiruzi 
The issue arises from using an incorrect parameter when executing the forge command. The problematic parameter is `broadcast`. This parameter is primarily used with `forge script`, not with `forge create`.

As far as I am aware of there are two solutions to this:
1. Edit the MakeFile and modify the cmd from `forge create` to `forge script`
```
deploy:
	@forge script script/DeployOurToken.s.sol:DeployOurToken --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast
```

Or 

2. Create an account using `cast`. Remove the `broadcast` parameter from your original command and append `--constructor-args 1000000` to the end.
  2a. How to create an account using `cast`
```
cast wallet import testAccount --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
```
  - FYI: the  private key is anvil's test account (address 0's) private key 
  - Also, after executing the cmd it will ask you for a password. Enter whatever password you would like.
  - You can view your EOA's via `cast wallet list`
  - You can also remove the EOA from your keystores via `rm -rf ~./foundry/keystores/testAccount`

    2b. After successfully creating an account copy and paste the below cmd, it will prompt you for the password you just created
```
forge create src/OurToken.sol:OurToken --rpc-url http://localhost:8545 --account testAccount --constructor-args 1000000
```


If you have any follow up issues, feel free to  @ me

**Reply by EngrPips** [reply]
Impressive 🙇🏾‍♂️
